src/main/java/com/amazon/opendistroforelasticsearch/security/resolver/IndexResolverReplacer.java

```diff
@@ -324,18 +320,26 @@ public Resolved resolveRequest(final Object request) {
         final Resolved.Builder resolvedBuilder = new Resolved.Builder();
         final AtomicBoolean isIndicesRequest = new AtomicBoolean();
         getOrReplaceAllIndices(request, new IndicesProvider() {
+            // resolve cache helps us big time on bulk requests
+            final ConcurrentHashMap<IndexResolveKey, Resolved> cache = new ConcurrentHashMap<>();
```

Comment:
@DmitryOlshansky Possibly I miss something, is it necessary to use thread safe `Map` for the `cache`?

Reply:
Indeed I was thinking about it ever since I did this patch. What thrown me off was that atomic boolean, but no amount of reviewing code suggests any concurrency going on here.

It would be best to put boolean inside provider struct itself and pass a reference then check it. 

Obviously I expect great saving in heap by switching to plain old Hasap